### ISSUE TITLE
Abnormal Traing loss

### ISSUE THREAD

**Issue by andypinxinliu** [mk-runner/MLRG#4]
Hi, sorry to borther you again. When I am training the first stage:
running script/MIMIC-CXR/run_cxr_pt_v0906_fs.sh

I find out the encoder does not match, I am not quite sure if it is right or not, here is the logging I have got:

Some weights of the model checkpoint at microsoft/BiomedVLP-CXR-BERT-specialized were not used when initializing CXRBertModel: ['bert.encoder.layer.8..output.LayerNorm.bias', 'bert.encoder.layer.11.intermediate.dense.weight', 'bert.encoder.layer.10.attention.self.value.weight', 'bert.encoder.layer..9.attention.self.key.bias',

....

tention.output.LayerNorm.weight', 'bert.encoder.layer.9.attention.self.query.weight', 'bert.encoder.layer.7.attention.self.value.weight', 'bert.encod
er.layer.6.attention.output.dense.weight']                                                                                                          
- This IS expected if you are initializing CXRBertModel from the checkpoint of a model trained on another task or with another architecture (e.g. ini
tializing a BertForSequenceClassification model from a BertForPreTraining model).                                                                   
- This IS NOT expected if you are initializing CXRBertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).                                                                      

Some weights of CXRBertModel were not initialized from the model checkpoint at microsoft/BiomedVLP-CXR-BERT-specialized and are newly initialized be$
ause the shapes did not match:                                                                                                                      
- bert.embeddings.word_embeddings.weight: found shape torch.Size([30522, 768]) in the checkpoint and torch.Size([50266, 768]) in the model instantiated                                                                                                                                                  
- cls.predictions.bias: found shape torch.Size([30522]) in the checkpoint and torch.Size([50266]) in the model instantiated                         
- cls.predictions.decoder.weight: found shape torch.Size([30522, 768]) in the checkpoint and torch.Size([50266, 768]) in the model instantiated     
- cls.predictions.decoder.bias: found shape torch.Size([30522]) in the checkpoint and torch.Size([50266]) in the model instantiated                 

You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.                                      
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1

In addition, I find out the training loss will soon get to NaN:

Epoch 0:   0%|              | 25/7506 [06:17<31:23:45,  0.07it/s, v_num=2, train_step_instance_loss=nan.0, train_step_mpc_loss=nan.0, train_step_loss=nan.0]

I did not modify any code from the repo, may I know what could be the potential problems

**Comment by mk-runner** [comment]
Regarding the CXR-BERT model, we only utilized the first six layers, which resulted in some warnings during initialization.

I have re-downloaded the code from the repository, and it runs correctly without any abnormal training loss issues. To ensure consistency, I recommend using the same environment, particularly the same version of the transformers library.

Please let me know if you encounter any further issues.

Best regards,
Kang Liu

**Comment by JennDong** [comment]
What is the python version of the environment?

**Comment by mk-runner** [comment]
> What is the python version of the environment?

Python 3.9.0. 

For other environment configurations, please refer to `https://github.com/mk-runner/MLRG/blob/main/requirements.txt`.

**Comment by mk-runner** [comment]
> Hi! Thanks For the great work! I'm curious about the training resources you used for your models. What kind of GPUs did you use?

Thanks for your attention. 

- In Stage 1, we requested 36G of GPU memory.
- In Stage 2, we requested 40G of GPU memory.

**Comment by andypinxinliu** [comment]
Hi, sorry for a late reply. I checked the codel later and find out it seems like for Pretraining, the released code utilize tokenizer of GPT2 for CXR-BERT encoder, i tried to change it to the original CXR-BERT tokenizer and find out the Nan problem will be resolved. Can you please check if the current implementation is correct or not.

I am also wondering how much time it takes to finish the training for Stage 1 and Stage2. From the logging i see for MIMIC-CXR, you trained for 23 epoches and the best validation epoch is 8. How much time does it take to finish the whole process.

For Stage 2, it trained for 40 epoches and the best is 25. I am assuming you have used the 8 epoch from Stage 1 for Stage 2 training and select 25 as the final public result right?

**Comment by andypinxinliu** [comment]
Yeah, I then tried inference using script/MIMIC-CXR/run_cxr_test_mlrg_v1011.sh. I find out the inference result of the generated report will be:

{'F1-Radgraph-partial': 0.18428996643805218, 'chexbert_5_micro_f1': 0.0, 'chexbert_5_macro_f1': 0.0, 'chexbert_all_micro_p': 0.06682966585167074, 'chexbert_all_micro_r': 0.023692574400462293, 'chexbert_all_micro_f1': 0.03498293515358362, 'chexbert_all_macro_p': 0.004773547560833625, 'chexbert_all_macro_r': 0.07000569151963575, 'chexbert_all_macro_f1': 0.00893765441069612, 'BLEU_1': 0.16874589366594925, 'BLEU_2': 0.09628696259269588, 'BLEU_3': 0.06085555853691924, 'BLEU_4': 0.041860426212079874, 'METEOR': 0.08978380804864469, 'ROUGE_L': 0.22138419964464592, 'CIDer': 0.04597718284500625}


However, if i run the inference by the generated reports from the huggingface, it is:

{'F1-Radgraph-partial': 0.29074779642289006, 'chexbert_5_micro_f1': 0.5503549017590784, 'chexbert_5_macro_f1': 0.4862237881570195, 'chexbert_all_micro_p': 0.5489597467209407, 'chexbert_all_micro_r': 0.467591254935953, 'chexbert_all_micro_f1': 0.5050189837208092, 'chexbert_all_macro_p': 0.4399492462801775, 'chexbert_all_macro_r': 0.354060820803069, 'chexbert_all_macro_f1': 0.3641635446370755, 'BLEU_1': 0.41114996799739173, 'BLEU_2': 0.2769778918508422, 'BLEU_3': 0.20362264525354418, 'BLEU_4': 0.1582088781713785, 'METEOR': 0.17633810974007486, 'ROUGE_L': 0.3195399064699496, 'CIDer': 0.3599887171235285}

I think there might be some problem with the tokenizer for the CXR-BERT? But I am not 100% sure

**Comment by mk-runner** [comment]
Thank you for your detailed message and for sharing your observations. Below are the responses to your questions:

### **1. Tokenizer Implementation**
Our implementation is correct, though it may not be optimal.  Given that CXR-BERT and DistilGPT2 use distinct tokenizers with different vocabulary sizes, we had to make a trade-off:  

- **Option 1:** Use both tokenizers (with CXR-BERT frozen). However, this may lead to misalignment in `input_ids` interpretations between the models, though fine-tuning could partially address this issue.  
- **Option 2:** Unify the tokenizers to ensure consistent handling of `input_ids`. This would require fine-tuning CXR-BERT, meaning its parameters couldn’t remain frozen (as done in MLRG’s implementation)..  

We welcome further discussion on optimizing this design choice.

### **2. Training Time and Model Selection**
**Stage 1 (Pretraining):**  
- Total epochs: 23 (early stopping at 15).  
- Optimal epoch: 8 (determined by ${\cal {L}}_{pretrain}$ on the validation set).  
- Training time: ~3 days on a single GPU.  

**Stage 2 (Fine-tuning):**  
- Total epochs: 40 (early stopping at 15).  
- Optimal epoch: 25 (determined by **BLEU-4 + F1-CheXbert + F1-RadGraph** on the validation set).  
- Training time: ~4 days on a single GPU.  

**Released Models:**  
- Stage 1: Weights from epoch 8.  
- Stage 2: Weights from epoch 25 (initialized with Stage 1’s epoch 8 checkpoint).  

Let us know if you’d like additional details or clarifications. We’re happy to help!

Best regards,  
Kang Liu

**Comment by mk-runner** [comment]
I can confirm that the performance discrepancy is indeed caused by using different tokenizers. As we've observed, inconsistent tokenizers lead to misalignment in `input_ids` interpretations between model components.

To verify this, I reran the inference using our original code with the exact configuration specified in `script/MIMIC-CXR/run_cxr_test_mlrg_v1011.sh`. The results now perfectly match those reported in the paper, demonstrating that:

- The tokenizer implementation is crucial for model performance

- Using MLRG's original configuration (including the specified tokenizer) ensures full reproducibility

Best regards,
Kang Liu

**Comment by andypinxinliu** [comment]
Hi, really appreciate your prompt feedback. However, i still encounter some issues for either training the model and inference. Is it possible to schedule a personal online meeting ｡ﾟ･ (>﹏<) ･ﾟ｡，I will try to send a personal email later. Really appreciate the help!

**Comment by mk-runner** [comment]
Thank you for your interest in our work and for reaching out!

I’d be happy to schedule an online meeting to discuss the challenges you're facing with model training and inference. Your feedback is invaluable, and I’d be glad to assist further.

Feel free to email me at `kangliu422@gmail.com` with your availability and preferred meeting platform (e.g., Tencent Meeting or Zoom). Looking forward to our discussion!

Best regards,
Kang Liu
